components/ws-manager-mk2/controllers/timeout_controller_test.go

```diff
@@ -0,0 +1,242 @@
+// Copyright (c) 2022 Gitpod GmbH. All rights reserved.
+// Licensed under the GNU Affero General Public License (AGPL).
+// See License-AGPL.txt in the project root for license information.
+
+package controllers
+
+import (
+	"time"
+
+	wsk8s "github.com/gitpod-io/gitpod/common-go/kubernetes"
+	"github.com/gitpod-io/gitpod/ws-manager-mk2/pkg/activity"
+	workspacev1 "github.com/gitpod-io/gitpod/ws-manager/api/crd/v1"
+	"github.com/google/uuid"
+	. "github.com/onsi/ginkgo/v2"
+	. "github.com/onsi/gomega"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/utils/pointer"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/client/fake"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	// . "github.com/onsi/ginkgo/extensions/table"
+)
+
+var _ = Describe("TimeoutController", func() {
+	Context("timeouts", func() {
+		var (
+			now        = time.Now()
+			conf       = newTestConfig()
+			r          *TimeoutReconciler
+			fakeClient client.Client
+		)
+		BeforeEach(func() {
+			var err error
+			// Use a fake client instead of the envtest's k8s client, such that we can add objects
+			// with custom CreationTimestamps and check timeout logic.
+			fakeClient = fake.NewClientBuilder().WithScheme(k8sClient.Scheme()).Build()
+			r, err = NewTimeoutReconciler(fakeClient, conf, &activity.WorkspaceActivity{})
+			Expect(err).ToNot(HaveOccurred())
+		})
+
+		type testCase struct {
+			phase             workspacev1.WorkspacePhase
+			lastActivityAgo   *time.Duration
+			age               time.Duration
+			customTimeout     *time.Duration
+			updateStatus      func(ws *workspacev1.Workspace)
+			controllerRestart time.Time
+			expectTimeout     bool
+		}
+		DescribeTable("workspace timeouts",
```

Comment:
This is `ginkgo`'s approach of table-testing. Useful here to tests different timeout scenarios